github_app.go

```diff
@@ -0,0 +1,231 @@
+	if s.AuthConfig.Github == nil || s.AuthConfig.Github.AppId == 0 {
+		return nil
+	}
+
+	key := s.Expansions[githubAppPrivateKey]
+	if key == "" {
+		return nil
+	}
+
+	return &githubAppAuth{
+		AppId:      s.AuthConfig.Github.AppId,
+		privateKey: []byte(key),
+	}
+}
+
+// CreateInstallationToken uses the owner/repo information to request an github app installation id
+// and uses that id to create an installation token.
+func (s *Settings) CreateInstallationToken(ctx context.Context, owner, repo string, opts *github.InstallationTokenOptions) (string, error) {
+	const (
+		maxDelay   = 10 * time.Second
+		minDelay   = time.Second
+		maxRetries = 5
+	)
+
+	if owner == "" || repo == "" {
+		return "", errors.New("no owner/repo specified to create installation token")
+	}
+	authFields := s.getGithubAppAuth()
+	if authFields == nil {
+		return "", errors.New("GitHub app is not configured in admin settings")
+	}
+
+	retryConf := utility.NewDefaultHTTPRetryConf()
+	retryConf.MaxDelay = maxDelay
+	retryConf.BaseDelay = minDelay
+	retryConf.MaxRetries = maxRetries
+
+	httpClient := utility.GetHTTPRetryableClient(retryConf)
+	defer utility.PutHTTPClient(httpClient)
+
+	key, err := jwt.ParseRSAPrivateKeyFromPEM(authFields.privateKey)
+	if err != nil {
+		return "", errors.Wrap(err, "parsing private key")
+	}
+
+	itr := ghinstallation.NewAppsTransportFromPrivateKey(httpClient.Transport, authFields.AppId, key)
+	httpClient.Transport = itr
+	client := github.NewClient(httpClient)
+
+	installationID, err := getInstallationID(ctx, owner, repo)
+	if err != nil {
+		return "", errors.Wrapf(err, "getting cached installation id for '%s/%s'", owner, repo)
+	}
+	if installationID == 0 {
+		installationID, resp, err := client.Apps.FindRepositoryInstallation(ctx, owner, repo)
+		if err != nil {
+			if resp != nil && resp.StatusCode == http.StatusNotFound {
+				return "", errors.Wrapf(gitHubAppNotInstalledError, "installation id for '%s/%s' not found", owner, repo)
+			}
+			grip.Debug(message.WrapError(err, message.Fields{
+				"message": "error finding installation id",
+				"owner":   owner,
+				"repo":    repo,
+				"appId":   authFields.AppId,
+				"ticket":  "EVG-19966",
+			}))
+			return "", errors.Wrapf(err, "finding installation id for '%s/%s'", owner, repo)
+		}
+		if installationID == nil {
+			return "", errors.Errorf("Installation id for '%s/%s' not found", owner, repo)
+		}
+
+		// Cache the installation ID for owner/repo.
+		installation := GitHubAppInstallation{
+			Owner:          owner,
+			Repo:           repo,
+			InstallationID: installationID.GetID(),
+		}
+		if err := installation.Upsert(ctx); err != nil {
+			return "", errors.Wrapf(err, "saving installation id for '%s/%s'", owner, repo)
+		}
+	}
+
+	token, _, err := client.Apps.CreateInstallationToken(ctx, installationID, opts)
+	if err != nil || token == nil {
+		return "", errors.Wrapf(err, "creating installation token for installation id: %d", installationID)
```

Comment:
reminder this is going to return no error if token is nil and err is nil

Reply:
Also "installation id '%d'"

Reply:
we do wrap IDs -- in this case the id happens to be an int but typically they are strings.

Reply:
same for below too